### PR TITLE
Recalculate input width when input-related props are changed

### PR DESCRIPTION
**Pull request by joekrill**
This recalculates the input width when `autoresize` is true and the `query`, `placeholder`, or `classNames` prop changes. Previously it only recalculated on the `query` prop changing, but changing the placeholder text or the any of the input classes could also effect the calculated width. Additionally, this _unsets_ the width style if the `autoresize` property is changed from `true` to `false`.

### CODE REVIEW

File: lib/Input.js [hunk 1]
```diff
@@ -37,7 +47,9 @@ class Input extends React.Component {
   }
 
   componentDidUpdate (prevProps) {
-    if (this.props.autoresize && prevProps.query !== this.props.query) {
+    if (prevProps.autoresize !== this.props.autoresize ||
+      (this.props.autoresize && RESIZE_PROPS.some((prop) => this.props[prop] !== prevProps[prop]))
+    ) {
```
Comment (i-like-robots) [inline lib/Input.js]: For simplicity perhaps remove this logic entirely. I'd bet recalculating the width is faster than this too!

**Comment by i-like-robots** [comment]
Thanks for this @joekrill - you're quite right about the edge cases.

File: lib/Input.js (hunk 1 above)
Comment (joekrill) [inline lib/Input.js]: That could lead to an infinite loop, though, because `setState` will get called every time `componentDidUpdate` is called, which will cause a re-render, which in turn causes `componentDidUpdate` to be called again, where we'd call `setState` and cause a re-render, etc...  

So taking that approach requires a `shouldComponentUpdate` implementation which should see if any of the underlying state or props _actually_ requires a re-render (or extend from `React.PureComponent` instead of `React.Component`, but then you need React 15.3, and the current dependency is 15.2).

So long story short, that will actually make this patch a bit more complicated than it currently is.

Comment (i-like-robots) [inline lib/Input.js]: Good point... How about if this condition is changed specifically for the case you describe?

```
componentDidUpdate (prevProps, prevState) {
  if (this.state.inputWidth === prevState.inputWidth) {
    this.updateInputWidth()
  }
}
```

My feeling is that there are many props and we want to trigger a re-render when all but one changes. Conversely there is only one state property for which we _don't_ want to re-render for the reasons you describe above.

Am I making any sense?

Comment (joekrill) [inline lib/Input.js]: I'm not sure I totally follow. It sounds like that's effectively what the `RESIZE_PROPS` const does -- it's only updating the input width when a prop that may effect the width calculation is updated. I'm not sure effect checking if the calculated `inputWidth` (on state)  in `componentDidUpdate` would have. Seems it could prevent the `inputWidth` from being updated in cases where it needs to be recalculated?

Comment (i-like-robots) [inline lib/Input.js]: My thinking was that the check would be slower than a few unnecessary recalcs with the overall benefit of simply having less logic to maintain.

I'm happy to ship this as-is TBH, apologies for being slow - I'm super busy ATM

**Comment by i-like-robots** [comment]
I'm so sorry for the delay on merging this... to be honest I totally forgot as I've been so busy! Thank you again.
